Use TreeBuilder for Infra and Cloud.

martinpovolny opened the pull request:
@mzazrivec, @dclarizio, @h-kataria : This is still a WIP but I welcome any feedback.

My plan is to finish these trees doing some minor cleanups in the TreeBuilder code. Then I'd like to do some more significant restructuring of the TreeBuilder code creating a clear distinction between tree structure, view-state (@sb) manipulation and presentation (details of JSON generated). Unwire all those tree-node-type, tree-type case statements and branching and more.

I want to limit the use of `@sb` especially. All the test conditions that depend on `x_tree` or `x_tree_node` should all test `@name`. The use of `@sb` should be basically limited to access to the `:open_nodes` key for each tree.

The `name`, `type` of a tree and `class` of the TreeBuilder seem to be redundant in most cases. Need to figure out the situations where they are not.

chessbyte commented on vmdb/app/controllers/application_controller/explorer.rb:
```diff
@@ -345,30 +345,30 @@ def x_build_node_id(object, pid = nil, options = {})
+  #  end
+
+  #  return kids                                         # Return the node's children
+  #end
```
@martinpovolny Why leave this in commented?

chessbyte commented on vmdb/app/controllers/mixins/vm_show_mixin.rb:
```diff
@@ -81,7 +81,11 @@ def set_active_elements_authorized_user(tree_name, accord_name, add_nodes, klass
   # Add the children of a node that is being expanded (autoloaded), called by generic tree_autoload method
   def tree_add_child_nodes(id)
-    return x_get_child_nodes_dynatree(x_active_tree, id)
+    #return x_get_child_nodes_dynatree(x_active_tree, id)
```
@martinpovolny Why leave this in commented?

chessbyte commented on vmdb/app/controllers/vm_cloud_controller.rb:
```diff
@@ -10,26 +10,30 @@ class VmCloudController < ApplicationController
-      build_vm_tree(:instances, :instances_tree)  # Build V&T tree
-      @trees.push("instances_tree")
+      # build_vm_tree(:instances, :instances_tree)  # Build V&T tree
+      # @trees.push("instances_tree")
```
@martinpovolny Why leave this in commented?

martinpovolny commented:
@chessbyte : I will remove all of the above. I am fighting code that's repeated on multiple places with slight modification so I am commenting it out to be able to quickly compare. All will be removed before I cleanup the PR and remove the [WIP].

martinpovolny commented:
non-TreeBuilder trees:

```
app/controllers/report_controller/schedules.rb:    tree_nodes = x_build_dynatree(x_tree(name))
app/controllers/report_controller/reports.rb:    tree_nodes = x_build_dynatree(x_tree(name))
app/controllers/report_controller/dashboards.rb:    tree_nodes = x_build_dynatree(x_tree(name))
app/controllers/report_controller/saved_reports.rb:    tree_nodes = x_build_dynatree(x_tree(name))
app/controllers/report_controller/menus.rb:    tree_nodes = x_build_dynatree(x_tree(name))
app/controllers/report_controller/widgets.rb:    tree_nodes = x_build_dynatree(x_tree(name))
app/controllers/miq_policy_controller/conditions.rb:    tree_nodes = x_build_dynatree(x_tree(name))
app/controllers/miq_policy_controller/alerts.rb:    tree_nodes = x_build_dynatree(x_tree(name))
app/controllers/miq_policy_controller/alert_profiles.rb:    tree_nodes = x_build_dynatree(x_tree(name))
app/controllers/miq_policy_controller/events.rb:    tree_nodes = x_build_dynatree(x_tree(name))
app/controllers/miq_policy_controller/policies.rb:    tree_nodes = x_build_dynatree(x_tree(name))
app/controllers/miq_policy_controller/miq_actions.rb:    tree_nodes = x_build_dynatree(x_tree(name))
app/controllers/miq_policy_controller/policy_profiles.rb:    tree_nodes = x_build_dynatree(x_tree(name))
app/controllers/report_controller.rb:    tree_nodes = x_build_dynatree(x_tree(name))
```

martinpovolny commented:
@mzazrivec : ping

h-kataria commented:
@martinpovolny i am getting an error when i click on a link "Show parent and child VMs" on the VM summary screen, i see that view has been deleted in this PR.

 Missing partial vm_common/vmtree with {:locale=>[:en], :formats=>[:js, "application/ecmascript", "application/x-ecmascript", :html, :text, :js, :css, :ics, :csv, :png, :jpeg, :gif, :bmp, :tiff, :mpeg, :xml, :rss, :atom, :yaml, :multipart_form, :url_encoded_form, :json, :pdf, :zip], :handlers=>[:erb, :builder, :jbuilder, :rjs, :haml]}. Searched in: * "/home/hkataria/dev/manageiq/vmdb/app/views" [vm_infra/show]

h-kataria commented:
@martinpovolny getting error when clicking on a VM on the right in Cloud/Instances -> Instances Accordion:

 Error text:
undefined method `x_get_tree_ems_kids' for #<TreeBuilderInstancesFilter:0x000000154d31d0> [vm_cloud/x_show] 

When i clicked on a VM in the right cell in "Instances By Provider" accordion, it display VM summary on the right but does not select/expand the nodes in the tree

h-kataria commented:
@martinpovolny trees are missing in both accordions in Workloads explorer, when clicking on a VM in the right cell it blows up with following error:
 Error text:
undefined local variable or method `prefix' for #<TreeBuilderImagesFilter:0x000000153fac68> [vm_or_template/x_show]

martinpovolny commented:
@h-kataria : thx, going to fix those and add specs to cover the cases.

Fryguy commented:
Seems you have a debugger binding.pry call in there...is this still WIP?

martinpovolny commented:
@Fryguy : fixing bugs found by @h-kataria, left the `pry` there last night...

martinpovolny commented:
@h-kataria - I stricked through the bugz that I was able to reproduce and fix. But I cannot get the last one. Can you, please, retry and make the steps more explicit? Or send me a full stack trace, please.

Thx!

martinpovolny commented:
@h-kataria : fixed the issue you reported, added one more simple test

the deprecation warning is to be removed in a fugure refactoring/cleanup when I modify all the child populating methods to have extra `parent` argument

martinpovolny commented:
@h-kataria : rebased again, clicked through stuff, previously added some more tests

h-kataria commented:
@first time when i go to Infra/Virtual Machines explorer/VM & Templates accordion on a root node -> Clicking on a VM in the right cell does not expand tree on the left, even tho i see VM summary screen on the right, tree is still collapsed with a root node selected. This is only happening when selecting a VM in the right cell the very first time, probably something is not getting initialized properly.

martinpovolny commented:
@h-kataria : I cannot reproduce the issue. Each time I click on the right I get the correct node selected on the left. If I log out and log in again it's consistently working.

Can you, please, retest?

Rebased again...

h-kataria commented:
@martinpovolny, see attached screeshot:
![screenshot from 2015-06-22 11 17 43](https://cloud.githubusercontent.com/assets/3450808/8285279/5da7a31e-18d0-11e5-9cea-9173d2d250f7.png)
All i did was signed out/sign back in, went to Infra/ Virtual Machines, and clicked on the VM in the right cell, that didn't expand the tree on left

h-kataria commented:
@martinpovolny i am getting error:
Error text:
ERROR: column reference "ems_id" is ambiguous LINE 1: ...dhat','TemplateVmware','TemplateMicrosoft')) AND (ems_id IS ... ^ [vm_infra/tree_select]

When clicking on Archived node in VMs & Templates tree

[----] I, [2015-06-22T12:09:22.956885 #12312:1211988]  INFO -- : Completed 200 OK in 10719ms (Views: 4606.0ms | ActiveRecord: 145.0ms)
[----] I, [2015-06-22T12:09:51.962476 #12312:1211988]  INFO -- : Started POST "/vm_infra/tree_select/?id=xx-arch" for 127.0.0.1 at 2015-06-22 12:09:51 -0400
[----] I, [2015-06-22T12:09:52.043683 #12312:1211988]  INFO -- : Processing by VmInfraController#tree_select as JS
[----] I, [2015-06-22T12:09:52.043865 #12312:1211988]  INFO -- :   Parameters: {"id"=>"xx-arch"}
[----] E, [2015-06-22T12:09:52.208518 #12312:1211988] ERROR -- : PG::Error: ERROR:  column reference "ems_id" is ambiguous
LINE 1: ...dhat','TemplateVmware','TemplateMicrosoft')) AND (ems_id IS ...
                                                             ^
: SELECT COUNT(DISTINCT "vms"."id") FROM "vms" LEFT OUTER JOIN "hosts" ON "hosts"."id" = "vms"."host_id" LEFT OUTER JOIN "storages" ON "storages"."id" = "vms"."storage_id" LEFT OUTER JOIN "ext_management_systems" ON "ext_management_systems"."id" = "vms"."ems_id" LEFT OUTER JOIN "snapshots" ON "snapshots"."vm_or_template_id" = "vms"."id" LEFT OUTER JOIN "compliances" ON "compliances"."resource_id" = "vms"."id" AND "compliances"."resource_type" = $1 LEFT OUTER JOIN "operating_systems" ON "operating_systems"."vm_or_template_id" = "vms"."id" LEFT OUTER JOIN "hardwares" ON "hardwares"."vm_or_template_id" = "vms"."id" LEFT OUTER JOIN "taggings" ON "taggings"."taggable_id" = "vms"."id" AND "taggings"."taggable_type" = $2 LEFT OUTER JOIN "tags" ON "tags"."id" = "taggings"."tag_id" WHERE (((vms.type IN ('VmMicrosoft','VmVmware','VmRedhat','VmXen','TemplateXen','TemplateRedhat','TemplateVmware','TemplateMicrosoft')) AND (ems_id IS NULL AND storage_id IS NULL)))
[----] F, [2015-06-22T12:09:52.209309 #12312:1211988] FATAL -- : Error caught: [PG::Error] ERROR:  column reference "ems_id" is ambiguous
LINE 1: ...dhat','TemplateVmware','TemplateMicrosoft')) AND (ems_id IS ...
                                                             ^

/home/hkataria/.rvm/gems/ruby-2.2.0/bundler/gems/rails-db89410f3d77/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:641:in `prepare'
/home/hkataria/.rvm/gems/ruby-2.2.0/bundler/gems/rails-db89410f3d77/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:641:in `prepare_statement'
/home/hkataria/.rvm/gems/ruby-2.2.0/bundler/gems/rails-db89410f3d77/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:600:in `exec_cache'
/home/hkataria/.rvm/gems/ruby-2.2.0/bundler/gems/rails-db89410f3d77/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb:589:in `execute_and_clear'
/home/hkataria/.rvm/gems/ruby-2.2.0/bundler/gems/rails-db89410f3d77/activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb:160:in `exec_query'
/home/hkataria/.rvm/gems/ruby-2.2.0/bundler/gems/rails-db89410f3d77/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:351:in `select'
/home/hkataria/.rvm/gems/ruby-2.2.0/bundler/gems/rails-db89410f3d77/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:32:in `select_all'
/home/hkataria/.rvm/gems/ruby-2.2.0/bundler/gems/rails-db89410f3d77/activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb:68:in `block in select_all'
/home/hkataria/.rvm/gems/ruby-2.2.0/bundler/gems/rails-db89410f3d77/activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb:83:in `cache_sql'
/home/hkataria/.rvm/gems/ruby-2.2.0/bundler/gems/rails-db89410f3d77/activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb:68:in `select_all'
/home/hkataria/.rvm/gems/ruby-2.2.0/bundler/gems/rails-db89410f3d77/activerecord/lib/active_record/relation/calculations.rb:264:in `execute_simple_calculation'
/home/hkataria/.rvm/gems/ruby-2.2.0/bundler/gems/rails-db89410f3d77/activerecord/lib/active_record/relation/calculations.rb:221:in `perform_calculation'
/home/hkataria/.rvm/gems/ruby-2.2.0/bundler/gems/rails-db89410f3d77/activerecord/lib/active_record/relation/calculations.rb:127:in `calculate'
/home/hkataria/dev/manageiq/vmdb/lib/extensions/ar_virtual.rb:372:in `calculate_with_virtual'
/home/hkataria/.rvm/gems/ruby-2.2.0/bundler/gems/activerecord-deprecated_finders-419f0397d5af/lib/active_record/deprecated_finders/relation.rb:84:in `calculate'
/home/hkataria/.rvm/gems/ruby-2.2.0/bundler/gems/rails-db89410f3d77/activerecord/lib/active_record/relation/calculations.rb:125:in `calculate'
/home/hkataria/dev/manageiq/vmdb/lib/extensions/ar_virtual.rb:372:in `calculate_with_virtual'
/home/hkataria/.rvm/gems/ruby-2.2.0/bundler/gems/activerecord-deprecated_finders-419f0397d5af/lib/active_record/deprecated_finders/relation.rb:84:in `calculate'
/home/hkataria/.rvm/gems/ruby-2.2.0/bundler/gems/rails-db89410f3d77/activerecord/lib/active_record/relation/calculations.rb:42:in `count'
/home/hkataria/dev/manageiq/vmdb/app/models/rbac.rb:155:in `compute_total_count'
/home/hkataria/dev/manageiq/vmdb/app/models/rbac.rb:186:in `find_targets_filtered_by_ids'
/home/hkataria/dev/manageiq/vmdb/app/models/rbac.rb:217:in `find_targets_with_direct_rbac'
/home/hkataria/dev/manageiq/vmdb/app/models/rbac.rb:242:in `find_targets_with_rbac'
/home/hkataria/dev/manageiq/vmdb/app/models/rbac.rb:408:in `search'
/home/hkataria/dev/manageiq/vmdb/app/models/miq_report/search.rb:107:in `paged_view_search'
/home/hkataria/dev/manageiq/vmdb/app/controllers/application_controller.rb:1944:in `get_view'
/home/hkataria/dev/manageiq/vmdb/app/controllers/application_controller/ci_processing.rb:953:in `process_show_list'
/home/hkataria/dev/manageiq/vmdb/app/controllers/vm_common.rb:1503:in `get_node_info'
/home/hkataria/dev/manageiq/vmdb/app/controllers/vm_common.rb:1554:in `replace_right_cell'
/home/hkataria/dev/manageiq/vmdb/app/controllers/vm_common.rb:1339:in `tree_select'

martinpovolny commented:
@h-kataria : oh, well, I did not change a line ;-) and .... all was green, when I left  :-( 

Now it seems there's some regression on master. Will take a look...

martinpovolny commented:
@h-kataria : on master you don't get this error?

h-kataria commented:
@martinpovolny switched my database, now clicking on a VM in right cell issue is no longer happening.

martinpovolny commented:
@h-kataria : rebased again and tested the "Archived" issue. Also cannot reproduce on my setup :-(

martinpovolny commented:
@h-kataria : seems the last issue is not related to my PR : https://github.com/ManageIQ/manageiq/pull/3236

Also thip PR includes helper function to make the spec in #3236 simplier -- https://github.com/ManageIQ/manageiq/pull/2075/files#diff-38f8d05afa33137efe154ac143ea9621R60

The failing travis also seems unrelated to this PR...

And most of all I am already pretty tired of rebasing this PR ;-)

So can we, please, merge this?

chessbyte commented:
@martinpovolny Should the `wip` label be removed from this PR?

martinpovolny commented:
@chessbyte : removed

h-kataria commented:
@martinpovolny tested UI again, looks good other than the issues that are being fixed in #3236

chessbyte commented:
@h-kataria so are you waiting on something/someone before merging?

h-kataria commented:
@chessbyte i was waiting for Travis to finish.

martinpovolny commented:
@h-kataria : rebased once again to include @kbrock's test. Now, please, merge this or shoot me!

chessbyte commented:
@martinpovolny Please rebase and repush now that #3236 is merged
